### PR TITLE
Add Dockerfile for building Yarn

### LINKED ISSUE #1903: Dockerfile/Vagrantfile for developing on Yarn
Inspired by (although somewhat tangential to) this Twitter thread: https://twitter.com/Daniel15/status/799044089035440141

We should add a Dockerfile or Vagrantfile for people that want to develop on Yarn itself, such that there's a consistent development environment and it's easy for people to get started.  There's already third-party Dockerfiles for using Yarn, but nothing for actually developing on Yarn itself.

Maybe Ubuntu or Debian based.

### PR DESCRIPTION
**Pull request by Daniel15**
**Summary**
Adds a Dockerfile for building Yarn. This is mainly for usage with CircleCI 2.0, however it could be used for development too. Inspired by @ryanwohara's version (https://github.com/yarnpkg/yarn/compare/master...ryanwohara:circleci-2.0), however I didn't use anything from his Dockerfile.

Closes #1903

**Test plan**
Ran `docker build -t yarn-dev .` to build an image, then `docker run -it -v c:\src\yarn:/yarn yarn-dev` to run it and mount `c:\src\yarn` to `/yarn`. Built Yarn and the build process mostly seemed to work. I still need to do a bit more testing though.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Dockerfile for building Yarn.
+
+#FROM debian:stretch-slim
```
Comment (bestander) [inline Dockerfile]: needed comment?

Comment (Daniel15) [inline Dockerfile]: Oops - I was originally trying to use Debian, but I found that some things worked better with Ubuntu for some reason. I'll remove this comment when I get a chance.

**Comment by Daniel15** [comment]
> Merge at will

Will do, I just need to fully test the entire build process and ensure everything works. Ubuntu just has Node.js 4.x, so I'll need to check if any part of the build requires a newer Node.js version (in which case I'll add the NodeSource repo)

**Comment by Daniel15** [comment]
Pushed to Docker Hub: https://hub.docker.com/r/yarnpkg/dev/

**Pull request merged**